File: kevoree-core/org.kevoree.kcl/src/main/scala/org/kevoree/kcl/KevoreeLocalLoader.scala
/**
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.kevoree.kcl

import org.xeustechnologies.jcl.ProxyClassLoader
import java.lang.Class
import java.io.{ByteArrayInputStream, InputStream}
import actors.{OutputChannel, DaemonActor}
import java.util.concurrent.{Semaphore, Callable}
import org.slf4j.LoggerFactory

/**
 * Created by IntelliJ IDEA.
 * User: duke
 * Date: 19/03/12
 * Time: 14:53
 */

class KevoreeLocalLoader(classpathResources: KevoreeLazyJarResources, kcl: KevoreeJarClassLoader) extends ProxyClassLoader /*with DaemonActor*/ {

  order = 1

  def loadClass(className: String, resolveIt: Boolean): Class[_] = {
    var result = kcl.getLoadedClass(className)
    if (result == null) {
      val bytes = kcl.loadClassBytes(className)
      if (bytes != null) {
        acquireLock(className)
        result = kcl.getLoadedClass(className)
        if (result == null) {
          result = kcl.internal_defineClass(className, bytes)
        }
        releaseLock(className)
      }
    }
    result
  }

  def loadResource(name: String): InputStream = {
    val arr = classpathResources.getResource(name)
    if (arr != null) {
      return new ByteArrayInputStream(arr)
    }
    null
  }

  class AcquireLockCallable(className: String) extends Callable[Semaphore] {
    def call(): Semaphore = {
      if (locked.containsKey(className)) {
        val tuple = locked.get(className)
        locked.put(className,(tuple._1,(tuple._2+1)))
        tuple._1
      } else {
        val obj = new Semaphore(0)
        locked.put(className, (obj,1))
        null //don't block first thread
      }
    }
  }


  def acquireLock(className: String) {
    val call = new AcquireLockCallable(className)
    try {
      val obj: Semaphore = KCLScheduler.getScheduler.submit(call).get()
      if (obj != null){
        logger.debug("Lock KCL to avoid concurrency {}",className)
        obj.acquire()
      }
    } catch {
      case ie: java.lang.InterruptedException =>
      case _@e => {
        logger.error("Error while sync " + className + " KCL thread : {}", Thread.currentThread().getName, e)
      }
    }
  }

  class ReleaseLockCallable(className: String) extends Runnable {
    def run() {
      if (locked.containsKey(className)) {
        val lobj = locked.get(className)
        if(lobj == 1){
          locked.remove(className)
        } else {
          lobj._1.release()
        }
      }
    }
  }

  private val logger = LoggerFactory.getLogger(this.getClass)

  def releaseLock(className: String) {
    try {
      val call = new ReleaseLockCallable(className)
      KCLScheduler.getScheduler.submit(call).get()
    } catch {
      case ie: java.lang.InterruptedException =>
      case _@e => {
        logger.error("Error while sync " + className + " KCL thread : {}", Thread.currentThread().getName, e)
      }
    }

  }

  private val locked = new java.util.HashMap[String, Tuple2[Semaphore,Int]]

}

File: kevoree-core/org.kevoree.kcl/src/main/java/org/kevoree/kcl/loader/KevoreeLocalLoader.kt
package org.kevoree.kcl.loader

import java.io.InputStream
import java.io.ByteArrayInputStream
import java.util.concurrent.Callable
import org.kevoree.kcl.KevoreeLazyJarResources
import org.kevoree.kcl.KevoreeJarClassLoader
import org.kevoree.kcl.KCLScheduler
import org.kevoree.log.Log

/**
 * Created by IntelliJ IDEA.
 * User: duke
 * Date: 19/03/12
 * Time: 14:53
 */

class KevoreeLocalLoader(val classpathResources: KevoreeLazyJarResources, val kcl: KevoreeJarClassLoader): ProxyClassLoader() {

    KevoreeLocalLoader(classpathResources: KevoreeLazyJarResources, kcl: KevoreeJarClassLoader){
        order = 1
    }


    public override fun loadResource(name: String?): InputStream? {
        if(name != null){
            val arr = classpathResources.getResource(name)
            if (arr != null) {
                return ByteArrayInputStream(arr)
            }
        }
        return null
    }

    public override fun loadClass(className: String?, resolveIt: Boolean): Class<out Any?>? {
        var result = kcl.getLoadedClass(className!!)
        if (result == null) {
            val bytes = kcl.loadClassBytes(className)
            if (bytes != null) {
                acquireLock(className!!)
                result = kcl.getLoadedClass(className)
                if (result == null) {
                    result = kcl.internal_defineClass(className, bytes)
                }
                releaseLock(className!!)
            }
        }
        return result
    }

    inner class AcquireLockCallable(val className: String): Callable<Any> {
        override fun call(): Any? {
            if (locked.containsKey(className)) {
                return locked.get(className)!!
            } else {
                locked.put(className, Object())
                return null //don't block first thread
            }
        }
    }

    fun acquireLock(className: String) {
        val call = AcquireLockCallable(className)
        try {
            val obj: Any? = KCLScheduler.getScheduler().submit(call).get()
            if (obj != null){
                (obj as java.lang.Object).wait()
            }
        } catch(ie: java.lang.InterruptedException) {
        }
        catch(e: Exception){

            if(Log.ERROR){
                Log.error("Error while sync " + className + " KCL thread : " + Thread.currentThread().getName(), e)
            }

        }
    }

    inner class ReleaseLockCallable(val className: String): Runnable {
        override fun run() {
            if (locked.containsKey(className)) {
                val lobj = locked.get(className)!!
                locked.remove(className)
                synchronized(lobj, {
                    (lobj as java.lang.Object).notifyAll()
                })
            }
        }
    }

    fun releaseLock(className: String) {
        try {
            val call = ReleaseLockCallable(className)
            KCLScheduler.getScheduler().submit(call).get()
        } catch(ie: java.lang.InterruptedException) {
        }
        catch (e: Exception) {
            if(Log.ERROR){
                Log.error("Error while sync " + className + " KCL thread : " + Thread.currentThread().getName(), e)
            }
        }
    }

    private val locked = java.util.HashMap<String, Any>()

}
